sources/packages/backend/libs/integrations/src/services/disbursement-schedule/e-cert-generation.service.ts

```diff
@@ -281,27 +306,109 @@ export class ECertGenerationService {
    * Calculate the effective value for every award. The result of this calculation
    * will be the value used to generate the e-Cert.
    * @param disbursements all disbursements that are part of one e-Cert.
+   * @param entityManager used to execute the commands in the same transaction.
    */
-  private calculateEffectiveValue(disbursements: ECertDisbursementSchedule[]) {
+  private async calculateEffectiveValue(
+    disbursements: ECertDisbursementSchedule[],
+    entityManager: EntityManager,
+  ): Promise<void> {
     for (const disbursement of disbursements) {
       for (const disbursementValue of disbursement.disbursementValues) {
-        if (this.shouldStopFunding(disbursement, disbursementValue)) {
+        if (this.shouldStopFullTimeBCFunding(disbursement, disbursementValue)) {
+          disbursementValue.restrictionAmountSubtracted =
+            disbursementValue.valueAmount -
+            (disbursementValue.disbursedAmountSubtracted ?? 0) -
+            (disbursementValue.overawardAmountSubtracted ?? 0);
           disbursementValue.effectiveAmount = 0;
+          disbursementValue.restrictionSubtracted = {
+            id: disbursement.restrictionId,
+          } as Restriction;
         } else {
           const effectiveValue =
             disbursementValue.valueAmount -
             (disbursementValue.disbursedAmountSubtracted ?? 0) -
             (disbursementValue.overawardAmountSubtracted ?? 0);
           disbursementValue.effectiveAmount = round(effectiveValue);
+
+          const application = disbursement.studentAssessment.application;
+          await this.checkLifeTimeMaximumAndAddStudentRestriction(
+            disbursementValue,
+            application,
+            entityManager,
+          );
         }
       }
     }
   }
 
+  /**
+   * Check if the student hits/exceeds the life time maximum for the BCSL Award.
+   * If they hits/exceeds the life time maximum, then the award is reduced so the
+   * student hits the exact maximum value and the student restriction
+   * {@link  RestrictionCode.BCLM} is placed for the student.
+   * @param disbursementValue disbursement value.
+   * @param application application related to the disbursement.
+   * @param entityManager used to execute the commands in the same transaction.
+   */
+  private async checkLifeTimeMaximumAndAddStudentRestriction(
+    disbursementValue: DisbursementValue,
+    application: Application,
+    entityManager: EntityManager,
+  ): Promise<void> {
+    if (disbursementValue.valueType === DisbursementValueType.BCLoan) {
+      const student = application.student;
+      const maxLifetimeBCLoanAmount =
+        application.programYear.maxLifetimeBCLoanAmount;
+      const auditUser = await this.systemUsersService.systemUser();
+
+      const totalLifeTimeAmount =
```

Comment:
Please take a look at the `getOverawardBalance`. The idea for the e-Cert is that we can execute the methods in bulk and process the result in memory. Is the current way, unless I am missing something, for every single e-cert record there will be other 2 queries executed to get the sum from SIMS + SFAS.